Update promptfoo package

roelandvh opened the pull request:
This PR attempts to fix the `promptfoo` package from failing to build.

Initial error:
```
...
error: builder for '/nix/store/zm2lfyhva9kmxajhch5xjm59ryhny66y-promptfoo-0.79.0.drv' failed with exit code 1;
       last 25 log lines:
       > node_modules/better-sqlite3/build/gyp-mac-tool: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/v0vy1c59f0vf71h64caw17hzymvg0zms-python3-3.12.10/bin/python3"
       > node_modules/nice-napi/build/gyp-mac-tool: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/v0vy1c59f0vf71h64caw17hzymvg0zms-python3-3.12.10/bin/python3"
       > Finished npmConfigHook
       > patchPhase completed in 42 seconds
       > Running phase: updateAutotoolsGnuConfigScriptsPhase
       > Running phase: configurePhase
       > no configure script, doing nothing
       > Running phase: buildPhase
       > no Makefile or custom buildPhase, doing nothing
       > Running phase: installPhase
       > Executing npmInstallHook
       >
       > up to date, audited 1927 packages in 3s
       >
       > 444 packages are looking for funding
       >   run `npm fund` for details
       >
       > found 0 vulnerabilities
       > Finished npmInstallHook
       > Running phase: fixupPhase
       > checking for references to /private/tmp/nix-build-promptfoo-0.79.0.drv-0/ in /nix/store/1wp25wzcig55f7xjmshymp5666gsxs64-promptfoo-0.79.0...
       > patching script interpreter paths in /nix/store/1wp25wzcig55f7xjmshymp5666gsxs64-promptfoo-0.79.0
       > ERROR: noBrokenSymlinks: the symlink /nix/store/1wp25wzcig55f7xjmshymp5666gsxs64-promptfoo-0.79.0/lib/node_modules/promptfoo/node_modules/promptfoo-docs points to a missing target: /nix/store/1wp25wzcig55f7xjmshymp5666gsxs64-promptfoo-0.79.0/lib/node_modules/promptfoo/site
       > ERROR: noBrokenSymlinks: the symlink /nix/store/1wp25wzcig55f7xjmshymp5666gsxs64-promptfoo-0.79.0/lib/node_modules/promptfoo/node_modules/nextui points to a missing target: /nix/store/1wp25wzcig55f7xjmshymp5666gsxs64-promptfoo-0.79.0/lib/node_modules/promptfoo/src/web/nextui
       > ERROR: noBrokenSymlinks: found 2 dangling symlinks, 0 reflexive symlinks and 0 unreadable symlinks
       For full logs, run 'nix log /nix/store/zm2lfyhva9kmxajhch5xjm59ryhny66y-promptfoo-0.79.0.drv'.
...
```

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

### Steps taken

#### 1. Upgrade `promptfoo` to newer version

The initial upgrade was done by running `nix-shell -p nix-update` and then inside that shell `nix-update promptfoo`. This updated the version number and various hashes.

#### 2. Fix build

See the commit messages for which error each change is trying to address.

After these changes I was able to successfully build the package using `nix-build -A promptfoo`.

After that I was able to install promptfoo using `nix-env -f . -iA promptfoo`.

Finally, I was able to run `promptfoo`:
```
% nix-env -f . -iA promptfoo
installing 'promptfoo-0.115.3'
building '/nix/store/lidkacp847l2pw7g2d877ivaahqqwq96-user-environment.drv'...
% type promptfoo
promptfoo is /Users/roeland/.nix-profile/bin/promptfoo
% promptfoo --version
0.115.3
```

I've also tested it by trying out `promptfoo eval` and `promptfoo view` on an existing `promptfooconfig.yaml`.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

roelandvh committed:
Update promptfoo package from 0.79.0 to 0.115.1

This was done by running `nix-shell -p nix-update` and then from inside that shell, running `nix-update promptfoo`.

roelandvh committed:
Prevent Playwright being downloaded during build of the package

This solves the following error:
```
...
error: builder for '/nix/store/10zk098dm6s6wxj73a3vfd6bqklywbq9-promptfoo-0.115.1.drv' failed with exit code 1;
       last 25 log lines:
       > npm error }
       > npm error Error: unable to get local issuer certificate
       > npm error     at TLSSocket.onConnectSecure (node:_tls_wrap:1679:34)
       > npm error     at TLSSocket.emit (node:events:518:28)
       > npm error     at TLSSocket._finishInit (node:_tls_wrap:1078:8)
       > npm error     at ssl.onhandshakedone (node:_tls_wrap:864:12) {
       > npm error   code: 'UNABLE_TO_GET_ISSUER_CERT_LOCALLY'
       > npm error }
       > npm error /private/tmp/nix-build-promptfoo-0.115.1.drv-0/source/node_modules/playwright-core/lib/server/registry/index.js:1096
       > npm error       throw new Error(`Failed to download ${title}, caused by
       > npm error             ^
       > npm error
       > npm error Error: Failed to download Chromium 136.0.7103.25 (playwright build v1169), caused by
       > npm error Error: Download failure, code=1
       > npm error     at ChildProcess.<anonymous> (/private/tmp/nix-build-promptfoo-0.115.1.drv-0/source/node_modules/playwright-core/lib/server/registry/browserFetcher.js:94:32)
       > npm error     at ChildProcess.emit (node:events:518:28)
       > npm error     at ChildProcess._handle.onexit (node:internal/child_process:293:12)
       > npm error     at /private/tmp/nix-build-promptfoo-0.115.1.drv-0/source/node_modules/playwright-core/lib/server/registry/index.js:1096:13
       > npm error     at async Registry._downloadExecutable (/private/tmp/nix-build-promptfoo-0.115.1.drv-0/source/node_modules/playwright-core/lib/server/registry/index.js:1095:5)
       > npm error     at async Registry.install (/private/tmp/nix-build-promptfoo-0.115.1.drv-0/source/node_modules/playwright-core/lib/server/registry/index.js:989:9)
       > npm error     at async installBrowsersForNpmInstall (/private/tmp/nix-build-promptfoo-0.115.1.drv-0/source/node_modules/playwright-core/lib/server/registry/index.js:1220:3)
       > npm error
       > npm error Node.js v22.14.0
       > npm error Log files were not written due to an error writing to the directory: /nix/store/k6gxiw6p3rfy2xwv1r2z1z5za4sr8r15-promptfoo-0.115.1-npm-deps/_logs
       > npm error You can rerun the command with `--loglevel=verbose` to see the logs in your terminal
       For full logs, run 'nix log /nix/store/10zk098dm6s6wxj73a3vfd6bqklywbq9-promptfoo-0.115.1.drv'.
```

roelandvh committed:
Remove broken symlinks after install and prior to packaging

This solves the following error:
```
...
error: builder for '/nix/store/jjn5jckyrbbf1lbl7sy3jylhkvckg6vl-promptfoo-0.115.1.drv' failed with exit code 1;
       last 25 log lines:
       > patching script interpreter paths in node_modules
       > node_modules/better-sqlite3/build/gyp-mac-tool: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/vfdk6q81hdjqjfiqz8f92hibdck3kmn6-python3-3.12.10/bin/python3"
       > Finished npmConfigHook
       > patchPhase completed in 38 seconds
       > Running phase: updateAutotoolsGnuConfigScriptsPhase
       > Running phase: configurePhase
       > no configure script, doing nothing
       > Running phase: buildPhase
       > no Makefile or custom buildPhase, doing nothing
       > Running phase: installPhase
       > Executing npmInstallHook
       >
       > up to date, audited 748 packages in 9s
       >
       > 87 packages are looking for funding
       >   run `npm fund` for details
       >
       > found 0 vulnerabilities
       > Finished npmInstallHook
       > Running phase: fixupPhase
       > checking for references to /private/tmp/nix-build-promptfoo-0.115.1.drv-0/ in /nix/store/p8ps9pmbm3mkgjh1z5y250wpdcwfwcrw-promptfoo-0.115.1...
       > patching script interpreter paths in /nix/store/p8ps9pmbm3mkgjh1z5y250wpdcwfwcrw-promptfoo-0.115.1
       > ERROR: noBrokenSymlinks: the symlink /nix/store/p8ps9pmbm3mkgjh1z5y250wpdcwfwcrw-promptfoo-0.115.1/lib/node_modules/promptfoo/node_modules/app points to a missing target: /nix/store/p8ps9pmbm3mkgjh1z5y250wpdcwfwcrw-promptfoo-0.115.1/lib/node_modules/promptfoo/src/app
       > ERROR: noBrokenSymlinks: the symlink /nix/store/p8ps9pmbm3mkgjh1z5y250wpdcwfwcrw-promptfoo-0.115.1/lib/node_modules/promptfoo/node_modules/promptfoo-docs points to a missing target: /nix/store/p8ps9pmbm3mkgjh1z5y250wpdcwfwcrw-promptfoo-0.115.1/lib/node_modules/promptfoo/site
       > ERROR: noBrokenSymlinks: found 2 dangling symlinks, 0 reflexive symlinks and 0 unreadable symlinks
       For full logs, run 'nix log /nix/store/jjn5jckyrbbf1lbl7sy3jylhkvckg6vl-promptfoo-0.115.1.drv'.
```

joshheinrichs-shopify commented on pkgs/by-name/pr/promptfoo/package.nix:
```diff
@@ -27,4 +25,13 @@ buildNpmPackage rec {
   };
+
+   # Prevent Playwright download as that will cause the package build to fail
+  PLAYWRIGHT_SKIP_BROWSER_DOWNLOAD = "1";
```
i guess this is functionally equivalent to `env.PLAYWRIGHT_SKIP_BROWSER_DOWNLOAD = "1";` but i'd be nice to make that explicit for the sake of consistency since that's the form that's used elsewhere[1][2][3].

[1] https://github.com/NixOS/nixpkgs/blob/a8d832a56361df7fdeaa9774bceae8745d6403cd/pkgs/by-name/ma/marktext/package.nix#L125-L128
[2] https://github.com/NixOS/nixpkgs/blob/a8d832a56361df7fdeaa9774bceae8745d6403cd/pkgs/by-name/fi/fider/frontend.nix#L34-L36
[3] https://github.com/NixOS/nixpkgs/blob/a8d832a56361df7fdeaa9774bceae8745d6403cd/pkgs/by-name/an/antares/package.nix#L58-L59

joshheinrichs-shopify commented on pkgs/by-name/pr/promptfoo/package.nix:
```diff
@@ -27,4 +25,13 @@ buildNpmPackage rec {
     maintainers = [ maintainers.nathanielbrough ];
   };
+
+   # Prevent Playwright download as that will cause the package build to fail
```
is this a cursor-generated comment? in general it's nice to snip these if they're self-explanitory as they add noise.

joshheinrichs-shopify commented on pkgs/by-name/pr/promptfoo/package.nix:
```diff
@@ -27,4 +25,13 @@ buildNpmPackage rec {
+
+  # Remove broken symlinks after installation
+  postInstall = ''
+    rm -f $out/lib/node_modules/promptfoo/node_modules/app
+    rm -f $out/lib/node_modules/promptfoo/node_modules/promptfoo-docs
```
do we know what creates these symlinks / why these symlinks are broken? possibly they correspond to the workspaces in package.json[1][2]? i don't know the node ecosystem super well. it looks the broken symlink check was introduced in nixpkgs somewhat recently[3]. searching `rm $out/lib/node_modules` in nixpkgs i see a bunch of other packages that are affected by this hook but no explanation 😕 

[1] https://github.com/promptfoo/promptfoo/blob/a79652729595cf10b68253e3bce3d6d6d2ee75bd/package.json#L19-L22
[2] https://github.com/promptfoo/promptfoo/blob/a79652729595cf10b68253e3bce3d6d6d2ee75bd/site/package.json#L2
[3] https://github.com/NixOS/nixpkgs/pull/370750

i tried inspecting the homebrew promptfoo package and did not see either of those paths fwiw.

i don't think `-f` is required.

roelandvh replied:
I've removed both `-f`s (and the redundant comment).

I also don't know why these symlinks are created/broken. But with these `rm ...` in place I was able build and also run the resulting `promptfoo` version successfully for our use cases.

roelandvh replied:
> i'm okay with landing this as-is in our private repo since it appears to work.

@joshheinrichs-shopify I understand the above to mean you'd prefer not to merge this PR into `Shopify/nixpkgs` given it's public, but instead merge it into our own private `Shopify/nixpkgs-shopify`. It looks like the structure over there is slightly different. Would the following be the work involved to move this over to `Shopify/nixpkgs-shopify`?

1. Add `/pkgs/promptfoo/default.nix` with the exact same contents as `/pkgs/by-name/pr/promptfoo/package.nix` in this repo.
2. Add `promptfoo = pkgs.callPackage ./pkgs/promptfoo { };` to `default.nix`.

Would that cover it? Anything missing or wrong?

cc @vaidas-shopify in case you're interested.

vaidas-shopify replied:
> 1. Add `/pkgs/promptfoo/default.nix` with the exact same contents as `/pkgs/by-name/pr/promptfoo/package.nix` in this repo.
> 2. Add `promptfoo = pkgs.callPackage ./pkgs/promptfoo { };` to `default.nix`.
> 
> Would that cover it? Anything missing or wrong?

@roelandvh yes, this looks right

roelandvh replied:
I've created said PR [here](https://github.com/Shopify/nixpkgs-shopify/pull/11).

roelandvh committed:
Use `env = { ... }` for setting `PLAYWRIGHT_SKIP_BROWSER_DOWNLOAD = 1`

Also, remove redundant comment.

roelandvh committed:
Remove `-f` from `rm` in `postInstall`

Also, remove redundant comment.

vaidas-shopify reviewed (changes requested):
1. Let's move updated package to https://github.com/Shopify/nixpkgs-shopify
2. In the mean time it would be great to open PR on the nixpkgs upstream

roelandvh committed:
Removing `dontNpmBuild` to ensure promptfoo is installed properly

Without this, promptfoo would fail to run with this error:
```
% promptfoo --version
node:internal/modules/cjs/loader:1404
  throw err;
  ^

Error: Cannot find module '/nix/store/hxl79bpga73qz4z6pcxd903b6b0xbnrb-promptfoo-0.115.3/lib/node_modules/promptfoo/dist/src/main.js'
    at Function._resolveFilename (node:internal/modules/cjs/loader:1401:15)
    at defaultResolveImpl (node:internal/modules/cjs/loader:1057:19)
    at resolveForCJSWithHooks (node:internal/modules/cjs/loader:1062:22)
    at Function._load (node:internal/modules/cjs/loader:1211:37)
    at TracingChannel.traceSync (node:diagnostics_channel:322:14)
    at wrapModuleLoad (node:internal/modules/cjs/loader:235:24)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:171:5)
    at node:internal/main/run_main_module:36:49 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}
```

roelandvh committed:
Update promptfoo to version 0.115.3

This was done by running `nix-shell -p nix-update` and then from inside that shell, running `nix-update promptfoo`.

roelandvh commented:
Closing this PR in favor of https://github.com/Shopify/nixpkgs-shopify/pull/11 .
